src/derivative_utils.jl

```diff
@@ -54,10 +54,161 @@ function calc_J!(integrator, cache::OrdinaryDiffEqMutableCache, is_compos)
   is_compos && (integrator.eigen_est = opnorm(J, Inf))
 end
 
+"""
+    WOperator(mass_matrix,gamma,J[;transform=false])
+
+A linear operator that represents the W matrix of an ODEProblem, defined as
+
+```math
+W = MM - \\gamma J
+```
+
+or, if `transform=true`:
+
+```math
+W = \\frac{1}{\\gamma}MM - J
+```
+
+where `MM` is the mass matrix (a regular `AbstractMatrix` or a `UniformScaling`),
+`γ` is a real number proportional to the time step, and `J` is the Jacobian
+operator (must be a `AbstractDiffEqLinearOperator`). A `WOperator` can also be
+constructed using a `*DEFunction` directly as
+
+    WOperator(f,gamma[;transform=false])
+
+`f` needs to have a jacobian and `jac_prototype`, but the prototype does not need
+to be a diffeq operator --- it will automatically be converted to one.
+
+`WOperator` supports lazy `*` and `mul!` operations, the latter utilizing an
+internal cache (can be specified in the constructor; default to regular `Vector`).
+It supports all of `AbstractDiffEqLinearOperator`'s interface.
+"""
+mutable struct WOperator{T,
+  MType <: Union{UniformScaling,AbstractMatrix},
+  GType <: Real,
+  JType <: DiffEqBase.AbstractDiffEqLinearOperator{T}
+  } <: DiffEqBase.AbstractDiffEqLinearOperator{T}
+  mass_matrix::MType
+  gamma::GType
+  J::JType
+  transform::Bool       # true => W = mm/gamma - J; false => W = mm - gamma*J
+  _func_cache           # cache used in `mul!`
+  _concrete_form         # non-lazy form (matrix/number) of the operator
+  WOperator(mass_matrix, gamma, J; transform=false) = new{eltype(J),typeof(mass_matrix),
+    typeof(gamma),typeof(J)}(mass_matrix,gamma,J,transform,nothing,nothing)
+end
+function WOperator(f::DiffEqBase.AbstractODEFunction, gamma; transform=false)
+  @assert DiffEqBase.has_jac(f) "f needs to have an associated jacobian"
+  if isa(f, Union{SplitFunction, DynamicalODEFunction})
+    error("WOperator does not support $(typeof(f)) yet")
+  end
+  # Convert mass matrix, if needed
```

Comment:
That's fine for now. I am hoping to support time-dependent mass matrices but we can just add an issue that it's a feature to implement in the future.